src/data/power_series.lean

```diff
@@ -0,0 +1,1012 @@
+
+@[simp] lemma coeff_add (n : ℕ) (φ ψ : power_series α) :
+  coeff n (φ + ψ) = coeff n φ + coeff n ψ := rfl
+
+@[simp] lemma monomial_add (n : ℕ) (a b : α) :
+  (monomial n (a + b) : power_series α) = monomial n a + monomial n b :=
+mv_power_series.monomial_add _ _ _
+
+@[simp] lemma C_add (a b : α) : (C (a + b) : power_series α) = C a + C b :=
+monomial_add 0 a b
+
+lemma coeff_mul (n : ℕ) (φ ψ : power_series α) :
+  coeff n (φ * ψ) = (finset.nat.antidiagonal n).sum (λ p, coeff p.1 φ * coeff p.2 ψ) :=
+begin
+  symmetry,
+  apply finset.sum_bij (λ (p : ℕ × ℕ) h, (single () p.1, single () p.2)),
+  { rintros ⟨i,j⟩ hij, rw finset.nat.mem_antidiagonal at hij,
+    rw [finsupp.mem_antidiagonal_support, ← finsupp.single_add, hij] },
+  { rintros ⟨i,j⟩ hij, refl },
+  { rintros ⟨i,j⟩ ⟨k,l⟩ hij hkl,
+    simpa only [prod.mk.inj_iff, finsupp.unique_single_eq_iff] using id },
+  { rintros ⟨f,g⟩ hfg,
+    refine ⟨(f (), g ()), _, _⟩,
+    { rw finsupp.mem_antidiagonal_support at hfg,
+      rw [finset.nat.mem_antidiagonal, ← finsupp.add_apply, hfg, finsupp.single_eq_same] },
+    { rw prod.mk.inj_iff, dsimp,
+      exact ⟨finsupp.unique_single f, finsupp.unique_single g⟩ } }
+end
+
+@[simp] lemma C_mul (a b : α) : (C (a * b) : power_series α) = C a * C b :=
+mv_power_series.C_mul _ _
+
+instance C.is_semiring_hom : is_semiring_hom (C : α → power_series α) :=
+mv_power_series.C.is_semiring_hom
+
+instance coeff.is_add_monoid_hom (n : ℕ) : is_add_monoid_hom (coeff n : power_series α → α) :=
+{ map_zero := coeff_zero n,
+  map_add := coeff_add n }
+
+instance : semimodule α (power_series α) :=
+mv_power_series.semimodule
+
+section map
+variables {β : Type*} {γ : Type*} [comm_semiring β] [comm_semiring γ]
+variables (f : α → β) (g : β → γ)
+
+/-- The map between power series induced by a map on the coefficients.-/
+def map : power_series α → power_series β :=
+mv_power_series.map f
+
+@[simp] lemma map_id : (map (id : α → α) : power_series α → power_series α) = id := rfl
+
+lemma map_comp : (map (g ∘ f) : power_series α → power_series γ) = map g ∘ map f := rfl
+
+@[simp] lemma coeff_map (n : ℕ) (φ : power_series α) :
+  coeff n (map f φ) = f (coeff n φ) := rfl
+
+variables [is_semiring_hom f] [is_semiring_hom g]
+
+@[simp] lemma map_zero : map f (0 : power_series α) = 0 :=
+mv_power_series.map_zero f
+
+@[simp] lemma map_one : map f (1 : power_series α) = 1 :=
+mv_power_series.map_one f
+
+@[simp] lemma map_add (φ ψ : power_series α) : map f (φ + ψ) = map f φ + map f ψ :=
+mv_power_series.map_add f φ ψ
+
+@[simp] lemma map_mul (φ ψ : power_series α) : map f (φ * ψ) = map f φ * map f ψ :=
+mv_power_series.map_mul f φ ψ
+
+instance map.is_semiring_hom :
+  is_semiring_hom (map f : power_series α → power_series β) :=
+mv_power_series.map.is_semiring_hom f
+
+end map
+
+section trunc
+
+variables [decidable_eq α] (n : ℕ)
+
+/-- The `n`th truncation of a power series to a polynomial -/
+def trunc (φ : power_series α) : polynomial α :=
```

Comment:
I think this exists for normal polynomials under a different name. We should probably keep the names consistent.

Reply:
There might also be a case for `n` being a `with_bot nat`, as the inductions are often easier starting at the zero submodule.

Reply:
Do you mean `mod_by_monic`? I agree that it is similar. However, that one doesn't work in the multivariate case. I guess it depends on which emphasis we want to place. I was viewing this mostly as a way to get a (mv_)polynomial from a formal power series. That will be useful when doing limits (evaluating power series) and when proving completeness (in the topological or algebraic sense).

Reply:
I was wrong it turns out. I thought Kenny defined it, but he just defined the submodule of polynomials with degree le n